packages/@sanity/form-builder/src/inputs/PortableText/toolbar/Toolbar.tsx

```diff
@@ -0,0 +1,246 @@
+            disabled={disabled}
+            collapsed={collapsed}
+            groups={actionGroups}
+            isFullscreen={isFullscreen}
+          />
+        </ActionMenuBox>
+      )}
+
+      {showInsertMenu && (
+        <InsertMenuBox flex={collapsed ? undefined : 1} padding={isFullscreen ? 2 : 1}>
+          <InsertMenu
+            disabled={disabled}
+            collapsed={collapsed}
+            items={insertMenuItems}
+            isFullscreen={isFullscreen}
+          />
+        </InsertMenuBox>
+      )}
+
+      <FullscreenButtonBox padding={isFullscreen ? 2 : 1}>
+        <Tooltip
+          content={
+            <Flex align="center" padding={1}>
+              <Box flex={1} paddingX={1}>
+                <Text size={1}>{`${isFullscreen ? 'Collapse' : 'Expand'} editor`}</Text>
+              </Box>
+              <Box marginLeft={1}>
+                <Hotkeys keys={[`${IS_MAC ? 'Cmd' : 'Ctrl'}`, 'Enter']} />
+              </Box>
+            </Flex>
+          }
+          placement={isFullscreen ? 'bottom' : 'top'}
+          portal="default"
+        >
+          <Button
+            padding={2}
+            icon={isFullscreen ? CollapseIcon : ExpandIcon}
+            mode="bleed"
+            onClick={onToggleFullscreen}
+          />
+        </Tooltip>
+      </FullscreenButtonBox>
+    </RootFlex>
+  )
+})
+
+export function Toolbar(props: ToolbarProps) {
+  const {hotkeys, isFullscreen, readOnly, onFocus, onToggleFullscreen} = props
+  const features = useFeatures()
+  const editor = usePortableTextEditor()
+  const selection = usePortableTextEditorSelection()
+  const disabled = !selection
+
+  const {push} = useToast()
+
+  const resolveInitialValue = useCallback(
+    (type: Type) => {
+      let isSlow = false
+      const slowTimer = setTimeout(() => {
+        isSlow = true
+        push({
+          id: 'resolving-initial-value',
+          status: 'info',
+          title: 'Resolving initial value…',
+        })
+      }, SLOW_INITIAL_VALUE_LIMIT)
+      return resolveInitialValueForType((type as unknown) as SchemaType)
+        .then((value) => {
+          if (isSlow) {
+            // I found no way to close an existing toast, so this will replace the message in the
+            // "Resolving initial value…"-toast and then make sure it gets closed.
+            push({
+              id: 'resolving-initial-value',
+              status: 'info',
+              duration: 500,
+              title: 'Initial value resolved',
+            })
+          }
+          return value
+        })
+        .catch((error) => {
+          push({
+            title: `Could not resolve initial value`,
+            id: 'resolving-initial-value',
+            description: `Unable to resolve initial value for type: ${type.name}: ${error.message}.`,
+            status: 'error',
+          })
+          return undefined
+        })
+        .finally(() => clearTimeout(slowTimer))
+    },
+    [push]
+  )
```

Comment:
Would be great to extract this to a hook (e.g. `useInitialValue`).

Reply:
It reminds me, we should probably test this one!

Reply:
Gonna follow up this when we add more tests to the input.